streaming/base/dataset.py

```diff
@@ -367,201 +455,155 @@ def _get_progress(self, world: World) -> Tuple[int, int]:
-
-        # Each worker loads its slice of the sample ID tensor to iterate through.
-        # Tensor shape: (num nodes, ranks per node, workers per rank, samples per worker).
-        sample_id_nbytes = np.int64().nbytes
-        num_bytes = os.path.getsize(filename)
-        if num_bytes % sample_id_nbytes:
-            raise ValueError(f'Generated shuffle is invalid: {filename} ({num_bytes} bytes).')
-        num_samples = num_bytes // sample_id_nbytes
-        num_workers = world.num_nodes * world.ranks_per_node * world.workers_per_rank
-        if num_samples % num_workers:
-            raise ValueError(f'Generated shuffle is invalid: {filename} ({num_bytes} bytes).')
-        samples_per_worker = num_samples // num_workers
-        offset_in_bytes = world.worker * samples_per_worker * sample_id_nbytes
-        bytes_to_read = samples_per_worker * sample_id_nbytes
-        with open(filename, 'rb', 0) as fp:
-            fp.seek(offset_in_bytes)
-            data = fp.read(bytes_to_read)
-        sample_ids = np.frombuffer(data, np.int64)
-
-        # Wait for everyone to read their part.
-        self._worker_barrier(world.workers_per_node)
 
-        # Now clean up after ourselves.
-        if world.is_local_leader:
-            os.remove(filename)
-
-        return sample_ids
-
-    def _download_file(self, basename: str) -> str:
-        """Safely download a file from remote to local cache.
+        # Sample each shard of each stream according to their proportions/repeats/samples. This
+        # gives us the resampled size of each underlying shard, and a mapping from each fake
+        # "big" sample ID to its underlying "small" sample ID.
+        pick_per_shard, small_per_big = self._resample_streams(epoch)
+
+        # Partition the global sample space (of resampled "big" sample IDs) into a tensor of
+        # shape (num physical nodes, ranks per node, workers per rank, batches per worker,
+        # samples per batch) such that we have an elastically deterministic sample order.
+        big_ids = get_partitions(self.samples_per_epoch, self.num_canonical_nodes, world.num_nodes,
+                                 world.ranks_per_node, world.workers_per_rank, self.batch_size,
+                                 sample_in_epoch)
+
+        # If we need to shuffle, shuffle in a node-aware and *underlying* shard-aware way.
+        if self.shuffle:
+            shuffle = get_shuffle(self.shuffle_algo, pick_per_shard, self.num_canonical_nodes,
+                                  self.shuffle_seed, epoch)
+            big_ids = np.where(big_ids != -1, shuffle[big_ids], -1)
+
+        # Now that we have partitioning and shuffled with hallucinated "big" sample IDs, we
+        # don't need them anymore, and can convert back to underlying "small" sample IDs.
+        return np.where(big_ids != -1, small_per_big[big_ids], -1)
+
+    def _get_worker_samples(self, world: World, epoch: int,
```

Comment:
any particular reason you chose to switch to shared memory for this logic?

Reply:
- staying in memory is much faster than writing out potentially many GBs of data to disk and reading it back in parts
- this is especially the case when world size = 1
- users with slow disk have complained about this
- I originally went with the file method in the interest of getting things done simply and quickly, but have finally had time to do it right